datafusion/optimizer/src/common_subexpr_eliminate.rs

```diff
@@ -719,23 +741,10 @@ impl TreeNodeRewriter for CommonSubexprRewriter<'_> {
                         ));
                     }
 
-                    let (series_number, id) = &self.id_array[self.curr_index];
```

Comment:
This was one of the bugs. It uses an already incremented `self.curr_index`. Instead, we grab all of the values once (see new line 713), before all the `self.curr_index` shenanigans.